wdl-engine/src/eval.rs

```diff
@@ -678,139 +616,283 @@ impl InputTrieNode<'_> {
 ///
 /// From the root to a terminal node represents a unique input.
 #[derive(Debug)]
-pub struct InputTrie<'a> {
+pub struct InputTrie {
+    /// The guest inputs root directory.
+    ///
+    /// This is `None` for backends that don't use containers.
+    guest_inputs_dir: Option<&'static str>,
     /// The URL path children of the tree.
     ///
     /// The key in the map is the scheme of each URL.
-    ///
-    /// A `BTreeMap` is used here to get a consistent walk of the tree.
-    urls: BTreeMap<&'a str, InputTrieNode<'a>>,
+    urls: HashMap<String, InputTrieNode>,
     /// The local path children of the tree.
     ///
     /// The key in the map is the first component of each path.
-    ///
-    /// A `BTreeMap` is used here to get a consistent walk of the tree.
-    paths: BTreeMap<&'a str, InputTrieNode<'a>>,
+    paths: HashMap<String, InputTrieNode>,
+    /// The inputs in the trie.
+    inputs: Vec<Input>,
     /// The next node identifier.
     next_id: usize,
-    /// The number of inputs in the trie.
-    count: usize,
 }
 
-impl<'a> InputTrie<'a> {
+impl InputTrie {
+    /// Constructs a new inputs trie with the given guest inputs directory.
+    pub fn new(guest_inputs_dir: &'static str) -> Self {
+        Self {
+            guest_inputs_dir: Some(guest_inputs_dir),
+            ..Default::default()
+        }
+    }
```

Comment:
Having both a `new()` and a `default()` is ripe for confusion. If there are two primary configurations for new tries, I think they should each have constructors with descriptive names, and we should forego a `Default` implementation.